addons/website/static/src/js/text_processing.js

```diff
@@ -0,0 +1,345 @@
+ * - "free": draw the path shape using the template only.
+ * @param {Function} options.template Returns a list of SVG path
+ * commands adapted to the container's size.
+ * @returns {String[]}
+ */
+function drawPath(textEl, options) {
+    const {width, height} = textEl.getBoundingClientRect();
+    const yStart = options.position === "center" ? height / 2 : height;
+
+    switch (options.mode) {
+        case "pattern": {
+            let i = 0, d = [];
+            const nbrChars = textEl.textContent.length;
+            const w = width / nbrChars, h = height * 0.2;
+            while (i < nbrChars) {
+                d.push(options.template(w, h));
+                i++;
+            }
+            return [`M 0,${yStart} ${d.join(" ")}`];
+        }
+        case "line": {
+            return [`M 0,${yStart} h ${width}`];
+        }
+    }
+    return options.template(width, height);
+}
+
+/**
+ * Returns a new highlight SVG adapted to the text container.
+ *
+ * @param {HTMLElement} textEl
+ * @param {String} highlightID
+ */
+export function drawTextHighlightSVG(textEl, highlightID) {
+    const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
+    svg.setAttribute("fill", "none");
+    svg.classList.add(
+        "o_th_svg",
+        // Identifies DOM content that should not be merged by the editor, even
+        // on identical parents.
+        "o_content_no_merge",
+        "position-absolute",
+        "overflow-visible",
+        "top-0",
+        "start-0",
+        "w-100",
+        "h-100",
+        "pe-none");
+    _textHighlightFactory[highlightID](textEl).forEach(d => {
+        const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
+        path.classList.add(`o_th_path_${highlightID}`);
+        path.setAttribute("stroke-linecap", "round");
+        path.setAttribute("stroke", "var(--text-highlight-color)");
+        path.setAttribute("stroke-width", "var(--text-highlight-width)");
+        path.setAttribute("d", d);
+        svg.appendChild(path);
+    });
+    return svg;
+}
+
+/**
+ * Divides the content of a text container into multiple
+ * `.o_text_highlight_item` units, and applies the highlight
+ * on each unit.
+ *
+ * @param {HTMLElement} topTextEl
+ * @param {String} highlightID
+ */
+export function setTextHighlight(topTextEl, highlightID) {
+    const lines = [];
+    let lineIndex = 0;
+    const nodeIsBR = node => node.nodeName === "BR";
+    const isRTL = el => window.getComputedStyle(el).direction === "rtl";
+
+    [...topTextEl.childNodes].forEach(child => {
+        // We consider `<br/>` tags as full text lines to ease
+        // excluding them when the highlight is applied on the DOM.
+        if (nodeIsBR(child)) {
+            lines[++lineIndex] = [child];
+            return lineIndex++;
+        }
+        const textLines = splitNodeLines(child);
+
+        // Special case: The text lines detection code in `splitNodeLines()`
```

Comment:
Couldn't this method have been used for all cases?

Reply:
As discussed on discord, the `splitNodeLines()` does the global "line break" check, and the code in the "Special case" is only used to handle a situation where the line break is not detected by `splitNodeLines` (when a line exactly ends with the current child node...).